[GRPC] Documentation for 3.2 GRPC GA (Bulk and k-NN Search) (#10651)
opensearch-project/documentation-website: _api-reference/grpc-apis/index.md

**Introduced 3.0**
{: .label .label-purple }

**Bulk and k-NN search generally available 3.2**
{: .label .label-green }

Starting with OpenSearch version 3.2, the gRPC [Bulk API]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/bulk/) and [k-NN search queries]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/knn/) are generally available. These use [protobuf version 0.6.0](https://github.com/opensearch-project/opensearch-protobufs/releases/tag/0.6.0). However, expect updates to the protobuf structure as the feature matures in the next releases. Other gRPC search functionality remains experimental and not recommended for production use. For updates on the progress of features or to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/16787).

> {: .note}

Author: The protobufs would in generally be auto-converted from the API specification, so I don't see adding protobufs to be the major challenge in the future, but rather the implementation on the server side. Is opensearch-protobufs the right place to host this issue, or core?

> {: .note}

Author: But I agree it would be highly encouraged and super helpful if the community could chip in to help implementing server-side GRPC query types/other endpoints. I would propose we: 

1. Create issues in core with the `help-needed` label on what APIs or protobuf fields are open to help  on being implemented on server side (which already have the protobufs generated)
2. Continue to use https://github.com/opensearch-project/OpenSearch/issues/16787 in the public docs here, as it's the "main"/meta GRPC issue, where people can feel free to leave general suggestions on what they would like to see in GRPC 
3. Enhance the README in the opensearch-protobufs repo to encourage users to: 
     a) file issues to report their protobuf-related feedback; and 
     b) have preliminary protobufs generated for certain APIs, which they wish to help/see implemented on the server side 

What do you think?

The OpenSearch gRPC functionality provides an alternative, high-performance transport layer using [gRPC](https://grpc.io/) for communication with OpenSearch. It uses protocol buffers over gRPC for lower overhead and faster serialization. This reduces overhead, speeds up serialization, and improves request-side latency, based on initial benchmarking results.

The primary goal of gRPC support is to:

* Offer a **binary-encoded** alternative to HTTP/REST-based communication.
* **Improve performance** for bulk workloads and large-scale ingestion scenarios.
* **Enable more efficient client integrations** across languages, like Java, Go, and Python, using native gRPC stubs.

## Performance benefits

Using gRPC APIs provides several advantages over HTTP APIs:

- **Reduced latency**: Binary protocol buffers eliminate JSON parsing overhead.
- **Higher throughput**: More efficient network utilization for high-frequency queries
- **Lower CPU usage**: Reduced serialization and deserialization costs.
- **Type safety**: Protocol buffer schemas provide compile-time validation.
- **Smaller payload sizes**: Binary encoding reduces network traffic.

> ## Enabling gRPC APIs

Reviewer: This section is not displaying correctly when building locally. To fix, add a new line on line 42, move lines 42-53 all the way to the left (no margin), and remove `1.` on line 55.

Author: good catch, updated per your suggestions

{: .note}

The OpenSearch gRPC functionality provides an alternative, high-performance transport layer using [gRPC](https://grpc.io/) for communication with OpenSearch. It uses protocol buffers over gRPC for lower overhead and faster serialization. This reduces overhead, speeds up serialization, and improves request-side latency, based on initial benchmarking results.

The primary goal of gRPC support is to:

* Offer a **binary-encoded** alternative to HTTP/REST-based communication.
* **Improve performance** for bulk workloads and large-scale ingestion scenarios.
* **Enable more efficient client integrations** across languages, like Java, Go, and Python, using native gRPC stubs.

## Performance benefits

Using gRPC APIs provides several advantages over HTTP APIs:

- **Reduced latency**: Binary protocol buffers eliminate JSON parsing overhead

> - **Better throughput**: More efficient network utilization for high-frequency queries

Author: Changed to `Higher throughput`

**Introduced 3.0**
{: .label .label-purple }

**Bulk and k-NN search generally available 3.2**
{: .label .label-green }

> Starting with OpenSearch version 3.2, the gRPC [Bulk API]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/bulk/) and [k-NN search queries]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/knn/) are generally available. These use [protobuf version 0.6.0](https://github.com/opensearch-project/opensearch-protobufs/releases/tag/0.6.0). However, expect updates to the protobuf structure as the feature matures in the next releases. Other gRPC search functionality remains experimental and not recommended for production use. For updates on the progress of features or to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/16787).

Reviewer 2: ```suggestion
Starting with OpenSearch version 3.2, the gRPC [Bulk API]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/bulk/) and [k-NN search queries]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/knn/) are generally available. These use [protobuf version 0.6.0](https://github.com/opensearch-project/opensearch-protobufs/releases/tag/0.6.0). However, expect updates to the protobuf structure as the feature matures in upcoming versions. Other gRPC search functionality remains experimental and not recommended for production use. For updates on the progress of these features or to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/16787).
```

{: .note}

The OpenSearch gRPC functionality provides an alternative, high-performance transport layer using [gRPC](https://grpc.io/) for communication with OpenSearch. It uses protocol buffers over gRPC for lower overhead and faster serialization. This reduces overhead, speeds up serialization, and improves request-side latency, based on initial benchmarking results.

The primary goal of gRPC support is to:

* Offer a **binary-encoded** alternative to HTTP/REST-based communication.
* **Improve performance** for bulk workloads and large-scale ingestion scenarios.
* **Enable more efficient client integrations** across languages, like Java, Go, and Python, using native gRPC stubs.

## Performance benefits

Using gRPC APIs provides several advantages over HTTP APIs:

- **Reduced latency**: Binary protocol buffers eliminate JSON parsing overhead.

> - **Higher throughput**: More efficient network utilization for high-frequency queries

Reviewer 2: ```suggestion
- **Higher throughput**: More efficient network utilization for high-frequency queries.
```

grpc.publish_host: 10.74.124.163
grpc.bind_host: 0.0.0.0
```
{% include copy.html %}

## Advanced gRPC settings

OpenSearch supports the following advanced settings for gRPC communication. These settings can be configured in `opensearch.yml`.

| Setting name | Description | Example value | Default value |
| :---- | :---- | :---- | :---- |
| `grpc.publish_port` | The external port number that this node uses to publish itself to peers for gRPC transport. | `9400` | `-1` (disabled) |
| `grpc.host` | A list of addresses the gRPC server will bind to. | `["0.0.0.0"]` | `[]` |
| `grpc.bind_host` | A list of addresses to bind the gRPC server to. Can be distinct from publish hosts. | `["0.0.0.0", "::"]` | Value of `grpc.host` |
| `grpc.publish_host` | A list of hostnames or IPs published to peers for client connections. | `["thisnode.example.com"]` | Value of `grpc.host` |
| `grpc.netty.worker_count` | The number of Netty worker threads for the gRPC server. Controls concurrency and parallelism. | `2` | Number of processors |
| `grpc.netty.max_concurrent_connection_calls` | The maximum number of simultaneous in-flight requests allowed per client connection. | `200` | `100` |

> | `grpc.netty.max_connection_age` | The maximum age a connection can reach before being gracefully closed. Supports time units like `ms`, `s`, `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `500ms` | Not set (no limit) |

Reviewer 2: ```suggestion
| `grpc.netty.max_connection_age` | The maximum age a connection can reach before being gracefully closed. Supports time units like `ms`, `s`, or `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `500ms` | Not set (no limit) |
```

> | `grpc.netty.max_connection_idle` | The maximum duration a connection can be idle before being closed. Supports time units like `ms`, `s`, `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `2m` | Not set (no limit) |

Reviewer 2: ```suggestion
| `grpc.netty.max_connection_idle` | The maximum duration for which a connection can be idle before being closed. Supports time units like `ms`, `s`, or `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `2m` | Not set (no limit) |
```

| `grpc.netty.keepalive_timeout` | The amount of time to wait for `keepalive` ping acknowledgment before closing the connection. Supports [time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `1s` | Not set |

> | `grpc.netty.max_msg_size` | The maximum inbound message size for gRPC requests. Supports units like `b`, `kb`, `mb`, `gb`. See [Supported units]({{site.url}}{{site.baseurl}}/api-reference/units/). | `10mb` or `10485760` | `10mb` |

Reviewer 2: ```suggestion
| `grpc.netty.max_msg_size` | The maximum inbound message size for gRPC requests. Supports units like `b`, `kb`, `mb`, or `gb`. See [Supported units]({{site.url}}{{site.baseurl}}/api-reference/units/). | `10mb` or `10485760` | `10mb` |
```

| **grpc.netty.keepalive_timeout**                | Time to wait for `keepalive` ping acknowledgment before closing the connection. Supports time units.             | `1s`                  | Not set              |
| **grpc.netty.max_msg_size**                     | Maximum inbound message size for gRPC requests. Supports units like `b`, `kb`, `mb`, `gb`.                    | `10mb` or `10485760`  | `10mb`               |

### Example configuration

Here's an example of a complete gRPC configuration in `opensearch.yml`:

```yaml
# Basic gRPC transport configuration
aux.transport.types: [transport-grpc]
aux.transport.transport-grpc.port: '9400-9500'

# Advanced gRPC settings
grpc.host: ["0.0.0.0"]
grpc.bind_host: ["0.0.0.0", "::"]
grpc.publish_host: ["thisnode.example.com"]
grpc.publish_port: 9400
grpc.netty.worker_count: 4
grpc.netty.max_concurrent_connection_calls: 200
grpc.netty.max_connection_age: 500ms
grpc.netty.max_connection_idle: 2m
grpc.netty.keepalive_timeout: 1s
grpc.netty.max_msg_size: 10mb
```
{% include copy.html %}

### Notes

> - For duration-based settings (e.g., `max_connection_age`), you can use units such as `ms` (milliseconds), `s` (seconds), `m` (minutes).

Author: Removed the entire `## Notes` section

> - For size-based settings (e.g., `max_msg_size`), you can use units such as `b` (bytes), `kb`, `mb`, `gb`.

Reviewer: ```suggestion
```

> - All settings are node-scoped unless otherwise specified.

Reviewer: ```suggestion
All settings are node-scoped unless otherwise specified.
{: .note}
```

> These settings are similar to the [HTTP Network settings]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/network-settings/#advanced-http-settings) but specifically apply to gRPC communication. For more details about the `transport-grpc `module implementation, see the [OpenSearch `transport-grpc` module](https://github.com/karenyrx/OpenSearch/tree/protobufs-0.8.0-transport-grpc-fixes/modules/transport-grpc).

Reviewer: ```suggestion
These settings are similar to the [HTTP network settings]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/network-settings/#advanced-http-settings) but specifically apply to gRPC communication. For more information about the `transport-grpc `module implementation, see the [OpenSearch `transport-grpc` module](https://github.com/karenyrx/OpenSearch/tree/protobufs-0.8.0-transport-grpc-fixes/modules/transport-grpc).
```

Author: Removed
